### DISCUSSION TITLE
Understanding ``*_unordered_flat_map`` design

### DISCUSSION THREAD

**Discussion by DNKpp** [slavenf/sfl-library#15] [Q&A] [answered]
Hello,
I found this library today and, when looking at the naming schema, there are some obvious references to the common ``std::`` containers.

Let's talk about the ``small_unordered_flat_map``:
At first, I was very confused, because I expected a ``hash`` template-parameter.
At second, I was very confused, because the documentation states ``The complexity of insertion or removal of elements is O(1).`` but at the same time ``The complexity of search is O(N).``

I wondered, how an uniqueness check can be performed in ``O(1)``, when the search is already ``O(n)``.
So, I looked at the code and it's very obvious, that it's just a linear-search over all elements, which then branches out on whether the key was already present or not.

So, what's the point of that design? It seems very counter-intuitive when expecting something similar to ``std::unordered_map``.

**Comment by slavenf** [comment, marked as answer]
Hello.

Elements of `small_unordered_flat_map` (and similar containers) are always stored **contiguously** in the memory. That's the point. It is cache friendly, especially if you have small number of elements. Typical cache line on modern x64 processors about 64 bytes. 

I have used a lot these containers and performances have been significantly improved in my case (compared to std::unordered_xxx).

`small_unordered_flat_map` (and similar containers) is implemented as unsorted vector. That the reason why search is O(N) and insertion is O(1). I think I made mistake here: insertion actually requires search first, which is O(N), and insertion is at the end of vector, which is O(1). In total is O(N). I will fix this documentation mistake in next release.

**Comment by DNKpp** [comment]
Thanks for your answer.
Yeah, I do see that cache point, but I find the naming very misleading. Even the docs state, that it's similar to ``std::unordered_map``, which actually isn't the case. It's rather a plain ``std::vector`` with an extended interface. The only reason I can imagine to chose this container over a ``_flat_map`` is, when key isn't ordered and thus ``operator <`` does not work. But in that case I would rather use a hash-map. If that's also not possible, then I probably use the wrong key type^^

How every, did you think about an actual counterpart to ``std::unordered_map``, which utilizes hashing? Maybe called ``*_flat_hash_map`` or something like that.

**Reply by slavenf** [reply]
It is similar to `std::unordered_map` in the sense that it is **associative** container that contains **key-value** pairs and elements are **not sorted** in any particular order. 

Do you have a suggestion how to make batter that first line in documentation?

It could be discussed about name. `flat` part of name denotes that elements are contiguously in the memory. `unordered` part of name denotes that elements are not sorted.

Yes, I have plans to make actual counterpart to `std::unordered_map` which utilizes hashing. Naming convetion `*_flat_hash_map` seems good, but my idea was to call these containers `sfl::unordered_map`, `sfl::small_unordered_map` and `sfl::static_unordered_map`. Do you have any comments for this?

**Reply by slavenf** [reply]
BTW, I am currently working on sorted associative containers based on red-black tree: `sfl::map`/`set`, `sfl::small_map`/`set`, `sfl::static_map`/`set`. These would be *true* counterparts to `std::map`/`set`.

**Reply by DNKpp** [reply]
> It is similar to `std::unordered_map` in the sense that it is **associative** container that contains **key-value** pairs and elements are **not sorted** in any particular order.
> 
> Do you have a suggestion how to make batter that first line in documentation?
> 
Ok, I see. I wouldn't make a reference to ``std::unordered_map`` at all, as that has very specific implications on what users (or at least I) expect. The last statement 
> The requirements of [UnorderedAssociativeContainer](https://en.cppreference.com/w/cpp/named_req/UnorderedAssociativeContainer) are partionally met.

is then just more confusing, because literally the first statement on that cppreference page says: 
> Unordered associative containers are [Containers](https://en.cppreference.com/w/cpp/named_req/Container) that **provide fast lookup** of objects based on keys.

For me, that's one of the most important aspects of a associative container.

> It could be discussed about name. `flat` part of name denotes that elements are contiguously in the memory. `unordered` part of name denotes that elements are not sorted.

I would be fine with something like ``(small/static)_unique_vector``.

> Yes, I have plans to make actual counterpart to `std::unordered_map` which utilizes hashing. Naming convetion `*_flat_hash_map` seems good, but my idea was to call these containers `sfl::unordered_map`, `sfl::small_unordered_map` and `sfl::static_unordered_map`. Do you have any comments for this?
> BTW, I am currently working on sorted associative containers based on red-black tree: `sfl::map`/`set`, `sfl::small_map`/`set`, `sfl::static_map`/`set`. These would be _true_ counterparts to `std::map`/`set`.

I don't know, whether that's a good idea or not. I for myself like the ``stl``-containers. From my point of view, there is no need of re-implementing containers with the same properties. I would be interested in an actual flat version of ``std::unordered_map`` with underlying contiguous storage.
